Fix feed navigation shortcuts when first article is not 100% visible

gowtham2095 opened the pull request:
<!--
     For Work In Progress Pull Requests, please use the Draft PR feature,
     see https://github.blog/2019-02-14-introducing-draft-pull-requests/ for further details.

     For a timely review/response, please avoid force-pushing additional
     commits if your PR already received reviews or comments.

     Before submitting a Pull Request, please ensure you've done the following:
     - 📖 Read the Forem Contributing Guide: https://github.com/forem/forem/blob/master/CONTRIBUTING.md#create-a-pull-request.
     - 📖 Read the Forem Code of Conduct: https://github.com/forem/forem/blob/master/CODE_OF_CONDUCT.md.
     - 👷‍♀️ Create small PRs. In most cases this will be possible.
     - ✅ Provide tests for your changes.
     - 📝 Use descriptive commit messages.
     - 📗 Update any related documentation and include any relevant screenshots.

     NOTE: Pull Requests from forked repositories will need to be reviewed by
     a Forem Team member before any CI builds will run. Once your PR is approved
     with a `/ci` reply to the PR, it will be allowed to run subsequent builds without
     manual approval.
-->

## What type of PR is this? (check all applicable)

- [ ] Refactor
- [ ] Feature
- [x]  Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description
 The feature was introduced to navigate between articles using the keyboard commands j and k. Previously we were looking for the completely visible first element. There could be cases where we may have partially visible elements, but no completely visible element. 

## Related Tickets & Documents

Closes https://github.com/forem/forem/issues/11566

## QA Instructions, Screenshots, Recordings

### To reproduce

* Navigate to /top/infinity
* Make sure your browser's window is not big enough to fit the entire first article
* Try to initialize the article feed navigation (pressing j and k)
* Notice nothing is happening

### Expected behavior

When no article is focused on, the first article that is visible should be focused on.

We might want to have similar behavior to the current one: we look for the first article that is fully visible and then fall back to any that is partially visible if nothing is found. 

### UI accessibility concerns?

It helps improves accessibility by responding to keyboard events to navigate between articles 

## Added tests?

- [ ] Yes
- [ ] No, and this is why:
- [x] I need help with writing tests

## Added to documentation?

- [ ] [Developer Docs](https://docs.forem.com) and/or
      [Admin Guide](https://forem.gitbook.io/forem-admin-guide/)
- [ ] README
- [x] No documentation needed

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/forem/forem?pullRequest=11874) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/forem/forem?pullRequest=11874) before we can accept your contribution.<br/><hr/>**Hariharan** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/forem/forem?pullRequest=11874) it.</sub>

gowtham2095 commented:
https://github.com/forem/forem/issues/11566 Issue Discussion thread

reobin commented on app/javascript/utilities/viewport.js:
```diff
@@ -22,3 +22,37 @@ export function isInViewport(element, offsetTop = 0) {
+  const clientHeight =
+    window.innerHeight || document.documentElement.clientHeight;
+  const clientWidth = window.innerWidth || document.documentElement.clientWidth;
+  return (
```
There is some stuff that can be reused in all those conditions, and it will help make the condition be more readable overall.

I would consider creating variables for each of the bounding rectangle's sides that defines is that particular side is visible in the viewport.

For example:
```js
const topIsInViewport = boundingRect.top <= clientHeight && boundingRect.top >= offsetTop;
const rightIsInViewport = boundingRect.right >= 0 && boundingRect.right <= clientWidth;
const bottomIsInViewport = boundingRect.bottom >= offsetTop && boundingRect.bottom <= clientHeight;
const leftIsInViewport = boundingRect.left <= clientWidth && boundingRect.left >= 0;
```

That way, you can use those variables in any combination in order to get the value you want.

For example, if you want to know if an element is 100% in the viewport, then all of those variables need to be true:
```js
return (
  topIsInViewport &&
  bottomIsInViewport &&
  leftIsInViewport &&
  rightIsInViewport
);
```

And if you want to know if an element is partially in the viewport, it needs to satisfy the following 2 conditions:
- its top OR its bottom is visible
- its left OR its right is visible

That way, you can use the following condition:
```js
return (topIsInViewport || bottomIsInViewport) && (leftIsInViewport || rightIsInViewport);
```

With all of that said, I think we can merge our 2 viewport functions. We can pass it an optional parameter that defines if we want the element to be partially visible or completely visible. We can then use the corresponding condition in our return statement.

gowtham2095 replied:
Thank you so much for taking the time to review it and giving valuable suggestions @reobin. I have made the changes as per your suggestion. 

And also there was one more following case apart from the above-mentioned ones,

```
const topIsOutOfViewport = boundingRect.top <= offsetTop;
const bottomIsOutOfViewport = boundingRect.bottom >= clientHeight;
const elementSpansEntireViewport =
    topIsOutOfViewport && bottomIsOutOfViewport;
```

The _elementSpansEntireViewport_ should also be part of the partially visible section check as below

```
return (topIsInViewport || bottomIsInViewport || elementSpansEntireViewport) &&
        (leftIsInViewport || rightIsInViewport)
```

I have added the changes and pushed them. Review it when you find the time and let me know. Cheers!

gowtham2095 replied:
Hey @reobin!

I have made the changes you suggested. Can you go over it when you find the time and let me know if it is fine?

reobin replied:
Hey @gowtham2095!

Sorry for the delay on the review! I have been very busy with personal life stuff. I'll let the Forem team take care of this. Good work!

gowtham2095 replied:
Hi @reobin,

I understand. Not a problem. Take care!

juliannatetreault commented:
Hi, @gowtham2095! 👋  I just wanted to check in to see if you're still working on this PR. Please let us know if you have any questions or if you need any help!

gowtham2095 commented:
> Hi, @gowtham2095! 👋 I just wanted to check in to see if you're still working on this PR. Please let us know if you have any questions or if you need any help!

Hey @juliannatetreault!👋 Hope you are well. I guess it is done. It has to be reviewed by two people. I'm waiting for the green flag from the second reviewer. Not sure if I have missed out on something. If there is something that needs to be done to push it forward, kindly let me know.

juliannatetreault commented:
> Hey @juliannatetreault!👋 Hope you are well. I guess it is done. It has to be reviewed by two people. I'm waiting for the green flag from the second reviewer. Not sure if I have missed out on something. If there is something that needs to be done to push it forward, kindly let me know.

Hey, @gowtham2095! 

Thank you so much for the update! It looks like my awesome co-worker, @rhymes, added another reviewer to the PR, which should help move the review process along for you. 😊  If you have any further questions, please do not hesitate to reach out to me or the reviewers!

gowtham2095 commented:
> > Hey @juliannatetreault!👋 Hope you are well. I guess it is done. It has to be reviewed by two people. I'm waiting for the green flag from the second reviewer. Not sure if I have missed out on something. If there is something that needs to be done to push it forward, kindly let me know.
> 
> Hey, @gowtham2095!
> 
> Thank you so much for the update! It looks like my awesome team member, @rhymes, added another reviewer to the PR, which should help move the review process along for you. 😊 If you have any further questions, please do not hesitate to reach out to me or the reviewers!

That sounds great. Thank you @rhymes and @juliannatetreault :)

aitchiss commented on app/javascript/shared/components/useListNavigation.js:
```diff
@@ -211,15 +211,29 @@ function getPreviousElement(
   const elements = document.querySelectorAll(selector);
-  return Array.prototype.find.call(elements, (element) =>
-    isInViewport(element),
+  let completelyVisibleFirstElement = Array.prototype.find.call(
```
we don't change this so we can use `const`

```suggestion
  const completelyVisibleFirstElement = Array.prototype.find.call(
```

gowtham2095 replied:
Makes sense. Have addressed the same.

aitchiss commented on app/javascript/shared/components/useListNavigation.js:
```diff
@@ -211,15 +211,29 @@ function getPreviousElement(
+  if (completelyVisibleFirstElement) {
+    return completelyVisibleFirstElement;
+  }
+  let partiallyVisibleFirstElement = Array.prototype.find.call(
```
same here re: const

```suggestion
  const partiallyVisibleFirstElement = Array.prototype.find.call(
```

gowtham2095 replied:
Have addressed the same.

aitchiss commented on app/javascript/shared/components/useListNavigation.js:
```diff
@@ -211,15 +211,29 @@ function getPreviousElement(
 
 /**
- * Gets the first visible element that matches a selector
+ * Checks if the first completely visible element is present that matches a selector and returns if it is available
+ * If that isn't visible, it looks for the partially visible element that is present and returns that
+ * If no elements visible(like the banner case which could cover the entire viewport), we select the first element from the list
```
Nice! This is a really helpful summary

gowtham2095 replied:
Thank you! 😊

aitchiss commented on app/javascript/utilities/viewport.js:
```diff
@@ -7,18 +7,35 @@
  * @returns {boolean} isInViewport - true if the element is visible in the viewport
  */
-export function isInViewport(element, offsetTop = 0) {
+export function isInViewport(
+  element,
+  offsetTop = 0,
+  isPartiallyVisible = false,
+) {
```
(refactor/blocking) - the updates here seem to work great! Just a small change requested here to make the developer experience a bit easier now that we have three params - let's change it to expect them in an object that we destructure like so:

```suggestion
export function isInViewport({
  element,
  offsetTop = 0,
  acceptPartiallyVisible = false,
}) {
```

So when we call this function it's clearer what the params are, e.g. `isInViewport({element: myElement, acceptPartiallyVisible: true})`

gowtham2095 replied:
Yes, plus it gives us flexibility as well. And also a lot more clearer. 👍

aitchiss commented on app/javascript/utilities/viewport.js:
```diff
@@ -7,18 +7,35 @@
+  const bottomIsOutOfViewport = boundingRect.bottom >= clientHeight;
+  const elementSpansEntireViewport =
+    topIsOutOfViewport && bottomIsOutOfViewport;
+  return isPartiallyVisible
+    ? (topIsInViewport || bottomIsInViewport || elementSpansEntireViewport) &&
+        (leftIsInViewport || rightIsInViewport)
+    : topIsInViewport &&
+        bottomIsInViewport &&
+        leftIsInViewport &&
+        rightIsInViewport;
```
(nitpick / not-blocking) - I think in this case the ternary makes it a little harder to read as we have so many conditionals chained together. I'd suggest splitting out into an `if` to make it more readable:

```js
 if (allowPartialVisibility) {
    return (
      (topIsInViewport || bottomIsInViewport || elementSpansEntireViewport) &&
      (leftIsInViewport || rightIsInViewport)
    );
  }

  return (
    topIsInViewport &&
    bottomIsInViewport &&
    leftIsInViewport &&
    rightIsInViewport
  );
```

gowtham2095 replied:
Thanks! Yes, the ternary demanded more cognitive load. This looks a lot better.

aitchiss reviewed (changes requested):
Great to see this fix! I've given it a quick spin and it seems to be working great - just a few small suggestions here mostly to make the additions a bit more readable 👍

gowtham2095 commented:
> Great to see this fix! I've given it a quick spin and it seems to be working great - just a few small suggestions here mostly to make the additions a bit more readable 👍

Hey @aitchiss ,

Thanks for taking the time to provide valuable suggestions. I have addressed the review comments as a separate commit and pushed it back. Let me know if it is fine. Will squash them as a single commit and push it back.

aitchiss commented on app/javascript/utilities/viewport.js:
```diff
@@ -3,22 +3,45 @@
+  element,
+  offsetTop = 0,
+  allowPartialVisibility = false,
+} = {}) {
```
(refactor/blocking) I don't think we need the default empty object here - we don't want the function to be called without at least the `element` property, so we can safely expect an object passed in

```suggestion
}) {
```

gowtham2095 replied:
> Thanks for making those changes so quickly! Just one small thing left.
> 
> Don't forget to also sign the CLA linked above 🙂

@aitchiss Hey! 👋  I signed the CLA, but it isn't picking up. Not sure what is missed. Let me try again.

aitchiss reviewed (changes requested):
Thanks for making those changes so quickly! Just one small thing left.

Don't forget to also sign the CLA linked above 🙂

gowtham2095 commented:
> Thanks for making those changes so quickly! Just one small thing left.
> 
> Don't forget to also sign the CLA linked above 🙂

Hey, @aitchiss!  Have included that minor change and also signed CLA. Let me know if anything else needs to be covered 😊. Cheers!

aitchiss reviewed (approved):
Looks good to me! 🚀

citizen428 commented:
@nickytonline Pinging you here for a second approval for this PR, thanks!

msarit commented:
@forem/frontend Hey FrontEnd Team! 👋🏾  This PR needs one more review, thanks! 🙏🏾

nickytonline commented on app/javascript/utilities/viewport.js:
```diff
@@ -3,22 +3,45 @@
   const clientWidth = window.innerWidth || document.documentElement.clientWidth;
+  const topIsInViewport =
+    boundingRect.top <= clientHeight && boundingRect.top >= offsetTop;
+  const rightIsInViewport =
```
question (potentially blocking): Is it necessary to check if the right and left of the element is in the viewport? Shouldn't top and bottom be enough?

Also, since this is being revisited, we could probably leverage the [Intersection Observer API](https://developer.mozilla.org/en-US/docs/Web/API/Intersection_Observer_API)? Or maybe save that for a subsequent PR, cc: @aitchiss.

aitchiss replied:
Yeah I agree it would be nice to rework this to use intersection observer - I'd be happy to tackle in a later PR though, since the current implementation fixes the bug 👍

gowtham2095 replied:
Hey, @nickytonline! I have avoided using the Intersection observer since I felt it will be a little overhead. Intersection Observer will be listening to the viewport and callbacks will be triggered when articles come in and leave the viewport. I felt that is not what is required here. What we need is when 'j' and 'k' key events happen, we have to let the user select some element based on what is visible in the viewport. I think the requirement here is to select from what is present in the viewport rather than observing what comes in and leaves the viewport. So felt the Intersection API might not be necessary. Correct me if my understanding is wrong.

gowtham2095 replied:
And also regarding the _left_ and _right_ check. Since this "_isInViewport_" function is part of the utilities I had included the left and right check as well to fit its correct definition. If it is specific for only this case I can remove those checks as well.

nickytonline reviewed (commented):
One small change request and a question.

gowtham2095 commented on app/javascript/utilities/viewport.js:
```diff
@@ -3,22 +3,45 @@
- * isInViewport(element); // true or false
+ * isInViewport({element, allowPartialVisibility = true}); // true or false
  *
  * @param {object} element - The HTML element to check
```
@nickytonline Thanks for taking the time to review. I shall add the same and push.

aitchiss commented:
@nickytonline - any chance you're able to re-review now changes have been made?

nickytonline commented on app/javascript/shared/components/useListNavigation.js:
```diff
@@ -211,15 +211,29 @@ function getPreviousElement(
 function getFirstVisibleElement(selector) {
   const elements = document.querySelectorAll(selector);
-  return Array.prototype.find.call(elements, (element) =>
-    isInViewport(element),
+  const completelyVisibleFirstElement = Array.prototype.find.call(
+    elements,
+    (element) => isInViewport({ element }),
   );
+  if (completelyVisibleFirstElement) {
+    return completelyVisibleFirstElement;
+  }
+  const partiallyVisibleFirstElement = Array.prototype.find.call(
+    elements,
+    (element) => isInViewport({ element, allowPartialVisibility: true }),
+  );
+  if (partiallyVisibleFirstElement) {
+    return partiallyVisibleFirstElement;
+  }
+  return elements && elements[0];
```
Refactor (blocking): Instead of Array.prototype, use the `find` method on an actual array. This can be accomplished by [spreading](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Spread_syntax) the [NodeList](https://developer.mozilla.org/en-US/docs/Web/API/NodeList) into an array.

```suggestion
  const elements = [...document.querySelectorAll(selector)];
  const completelyVisibleFirstElement = elements.find((element) =>
    isInViewport({ element }),
  );

  if (completelyVisibleFirstElement) {
    return completelyVisibleFirstElement;
  }

  const partiallyVisibleFirstElement = elements.find((element) =>
    isInViewport({ element, allowPartialVisibility: true }),
  );

  if (partiallyVisibleFirstElement) {
    return partiallyVisibleFirstElement;
  }

  return elements[0];
}
```

gowtham2095 replied:
Sure @nickytonline. Thanks for taking the time to go over. Will push the suggested change.

gowtham2095 replied:
Have pushed it @nickytonline. Kindly check and let me know when you find time :)

nickytonline reviewed (changes requested):
Tested this locally @gowtham2095 and it works great. One small change request. Looking forward to seeing this in production! 🚀

nickytonline reviewed (approved):
Thanks so much for the PR @gowtham2095 and thanks for making those changes! Looking forward to seeing this in production. 🚢 😎
